uwblueprint/starter-code-v2
# Add download csv functionality from frontend utility

## Notion ticket link

Closes #68



## Implementation description
* Creating an anchor element and clicking it seems to be the most conventional way to download files, note that the anchor elements aren't actually added to the HTML body
* References: 
* material-ui : https://github.com/mui-org/material-ui-x/blob/fa346f0fbe3d9b9eea9bb403fe4675f544d6abf9/packages/grid/_modules_/grid/utils/exportAs.ts
* material-table : uses filefy https://github.com/mbrn/filefy/blob/master/src/core/BaseBuilder.ts



## Steps to test
1. Navigate to the Display Entities page
<img width="1283" alt="image" src="https://user-images.githubusercontent.com/37782734/127784875-bd115d51-099b-490a-b8b5-13c9dc52576e.png">


2. Click on the Download CSV button and verify that it downloads the csv with the entities in the table. This tests the GraphQL endpoint.
<img width="628" alt="image" src="https://user-images.githubusercontent.com/37782734/127784897-b4a9e7f8-e69e-4193-9633-b074c56b637f.png">

3. Uncomment the rest section of the downloadEntitiesCSV function (in DisplayTableContainer.tsx) and comment the GraphQL section. Update the imports accordingly so that the frontend compiles.
4. Repeat Step 2. This tests the REST endpoint.
5. Comment the rest section of the downloadEntitiesCSV function (in DisplayTableContainer.tsx) and uncomment the lines after the comment `Use the following lines to download CSV using frontend CSV generation instead of API`.
6. Repeat Step 2. This tests the frontend csv generation.


## What should reviewers focus on?
* Tested on Chrome and Safari (NOTE: on chrome I had a popup blocker so it prevented download unless I disabled it, I think this happens for other sites too though), needs to be checked on Firefox and possible IE
* This page mentions that Chrome disabled this manual link-clicking for cross-origin URLs, not sure how to test that though https://stackoverflow.com/questions/1066452/easiest-way-to-open-a-download-window-without-navigating-away-from-the-page

## Checklist
